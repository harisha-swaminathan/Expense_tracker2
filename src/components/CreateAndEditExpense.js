import React, { useState } from "react";
import DatePicker from "react-datepicker";

const CreateAndEditExpense = (props) => {
    const expenseToEdit = props.expenseToEdit || {};    
    const [description, setDescription] =  useState(expenseToEdit.description ? expenseToEdit.description : '');
    const [amount, setAmount] =  useState(expenseToEdit.amount ?  (expenseToEdit.amount/100).toString() : '');
    const [note, setNote] =  useState(expenseToEdit.note ? expenseToEdit.note : '');
    const [createdAt, setCreatedAt] =  useState(expenseToEdit.createdAt || new Date());
    const [error, setError] =  useState('');

    const onDescriptionChange = (e) => {
        setDescription(e.target.value)
    };
    const onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            setAmount(e.target.value);
        }
    };
    const onNoteChange = (e) => {
        setNote(e.target.value)
    };
    const onDateChange = (createdAt) => {
        if(createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onExpenseFormSubmit = (e) => {
        e.preventDefault();
        if(!description || !amount) {
            setError('Please provide Description and amount!');
            return;
        }
        setError('');
        props.onSubmit({
            description,
            amount: parseFloat(amount, 10) * 100,
            createdAt: createdAt.valueOf(),
            note
        });
    };
    return (
        <>
            {!!error && <p>{error}</p>}
            <form onSubmit={onExpenseFormSubmit} data-testid="form-element">
                <input
                    type="text"
                    placeholder="Description"
                    onChange={onDescriptionChange}
                    value = {description}
                    autoFocus
                />
                <input
                    type="text"
                    placeholder="Amount"
                    onChange={onAmountChange}
                    value = {amount}
                />
                <DatePicker 
                    selected={createdAt} 
                    showIcon
                    format="dd/MM/yy"
                    onChange={onDateChange} 
                />
                <textarea
                    placeholder="add a note for you expense"
                    onChange={onNoteChange}
                    value = {note}
                />
                <button data-testid="form-submit"> Add Expense</button>
            </form>
        </>
    );
}

export default CreateAndEditExpense;

