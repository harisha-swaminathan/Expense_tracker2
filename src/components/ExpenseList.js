import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
    return (
        <>
            {
                props.expenses.length ===0 ? (
                    <p> No expenses </p>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense}/>
                    })
                )}
        </>
    );
}
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;