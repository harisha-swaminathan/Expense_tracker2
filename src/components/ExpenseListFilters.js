import React, { useState } from 'react';
import {connect} from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate} from '../actions/filters';
import DatePicker from "react-datepicker";

export const ExpenseListFilter = (props) => {
    const startDate = props.filters.startDate ? props.filters.startDate.toString() : null;
    const endDate = props.filters.endDate ? props.filters.endDate.toString() : null;
    const text = props.filters.text;
    return (
        <>
        <input 
            placeholder='filter by text'
            type="text"
            value={text}
            onChange={(e) => {
                props.setTextFilter(e.target.value)
            }
        }/>
        <select
            data-testid="select-filter"
            value="date"
            onChange={(e) => {
                if(e.target.value === "amount"){
                    props.sortByAmount();
                } else {
                    props.sortByDate();
                }
            }
        }>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
        </select>
        <DatePicker
            selected={startDate}
            showIcon
            onChange={(date) => {
                props.setStartDate(date)
            }}
            selectsStart
            startDate={startDate}
            endDate={endDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        <DatePicker
            selected={endDate}
            showIcon
            onChange={(date) => {
                props.setEndDate(date)
            }}
            selectsEnd
            startDate={startDate}
            endDate={endDate}
            minDate={startDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        </>
    );
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => {
        dispatch(setTextFilter(value));
    },
    sortByAmount: () => {
        dispatch(sortByAmount());
    },
    sortByDate: () => {
        dispatch(sortByDate());
    },
    setStartDate: (date) => {
        dispatch(setStartDate(date));
    },
    setEndDate: (date) => {
        dispatch(setEndDate(date));
    }
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);