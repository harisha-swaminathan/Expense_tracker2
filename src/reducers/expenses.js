const expenseReducerDefaultState = [];

const expensesReducer = (state = expenseReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'DELETE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            })
        default:
            return state;
    }
};

export default expensesReducer;