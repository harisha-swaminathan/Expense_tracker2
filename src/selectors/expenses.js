import { isBefore, isAfter } from "date-fns";


const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
      
        const description = expense.description.toLowerCase();
        const textMatch = typeof text !== 'string' || description.includes(text.toLowerCase());
        const startDateMatch = !startDate || isAfter(expense.createdAt, startDate); 
        const endDateMatch =  !endDate || isBefore(expense.createdAt, endDate);

        return textMatch && startDateMatch && endDateMatch;
    }).sort((expense1, expense2) => {
        if(sortBy == 'date'){
            return expense1.createdAt < expense2.createdAt ? 1 : -1;
        } else {
            return expense1.amount < expense2.amount ? 1 : -1;
        }
    })
};

export default getVisibleExpenses;