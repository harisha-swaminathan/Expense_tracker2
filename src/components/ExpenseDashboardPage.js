import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const ExpensesDashboardPage = () => (
    <div>
        
        <ExpenseListFilter/>
        <ExpenseList/>
    </div>
);
export default ExpensesDashboardPage;