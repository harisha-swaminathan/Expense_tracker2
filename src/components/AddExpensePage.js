import React from 'react';
import CreateAndEditExpense from './CreateAndEditExpense';
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import { useNavigate } from "react-router";

export const AddExpensePage = (props) => {
    const navigator = useNavigate();
    return (
        <div>
            <h1> Add Expense</h1>
            <CreateAndEditExpense 
                onSubmit={(expense) => {
                   props.addExpense(expense)
                    navigator("/");
                }}
            />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    
    addExpense: ({description, note, amount, createdAt}) => {
        dispatch(addExpense({ 
            description,
            note,
            amount,
            createdAt
        }));
    }
    
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);