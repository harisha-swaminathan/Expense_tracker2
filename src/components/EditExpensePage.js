import React from 'react';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import CreateAndEditExpense from './CreateAndEditExpense';
import { editExpense , deleteExpense} from '../actions/expenses';
import { useNavigate } from "react-router";

export const EditExpensePage = (props) => {
    const navigator = useNavigate();
    const id = useParams().id;
    const expenseToEdit = props.expenses.find((expense) => expense.id === id);
    const onEditExpense = (updates) => {
        props.editExpense(id, updates);
        navigator("/");
    };
    const onRemove = () => {
        props.deleteExpense(id);
        navigator("/");
    };
    return (
        <div>
            <CreateAndEditExpense expenseToEdit={expenseToEdit} onSubmit={(onEditExpense)}/>
            <button onClick={onRemove}> Remove </button>
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        deleteExpense: (id) => {
            dispatch(deleteExpense(id))
        },
        editExpense: (id, updates) => {
            dispatch(editExpense(id, updates))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);