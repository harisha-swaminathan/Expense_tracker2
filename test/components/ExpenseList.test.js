import React from 'react';
import { render } from '@testing-library/react';
import { ExpenseList } from "../../src/components/ExpenseList";
import { BrowserRouter } from 'react-router-dom';
import expenses from '../fixtures/expenses';

test('expenseList is rendered with expenses as expected', () =>{
    const { asFragment } = render(<BrowserRouter> <ExpenseList expenses={expenses}/> </BrowserRouter>);
    expect(asFragment()).toMatchSnapshot();
});

test('expenseList is rendered with empty array as expected', () =>{
    const { asFragment } = render(<BrowserRouter> <ExpenseList expenses={[]}/> </BrowserRouter>);
    expect(asFragment()).toMatchSnapshot();
});