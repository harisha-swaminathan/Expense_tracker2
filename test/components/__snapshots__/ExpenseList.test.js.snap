// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expenseList is rendered with empty array as expected 1`] = `
<DocumentFragment>
   
  <p>
     No expenses 
  </p>
   
</DocumentFragment>
`;

exports[`expenseList is rendered with expenses as expected 1`] = `
<DocumentFragment>
   
  <a
    href="/edit/1"
  >
    <h3>
      celine bag
    </h3>
  </a>
   
  <br />
  <p>
    850 - 123
  </p>
  <a
    href="/edit/2"
  >
    <h3>
      diamond ring
    </h3>
  </a>
   
  <br />
  <p>
    1250 - 456
  </p>
  <a
    href="/edit/3"
  >
    <h3>
      jeans
    </h3>
  </a>
   
  <br />
  <p>
    80 - 789
  </p>
   
</DocumentFragment>
`;
